File: utilities/write_sam_from_mapped_reads_db_and_trace.py
#!/usr/bin/python

import sys
import os

# add python_lib to sys.path
sys.path.insert(0, os.path.normpath( sys.path[0] + "/../python_lib") )
from sam import *
from trace import *
from utils import *

def usage():
    print "Usage: ./mapped_reads_to_sam output_directory input.bin.trace [--use-nc]"
    sys.exit(1)

def parse_args():
    import argparse; parser = argparse.ArgumentParser()

    # Required arguments
    parser.add_argument("output_directory",
        help="Path to Statmap output directory to process")
    parser.add_argument("input_trace",
        help="Binary trace file to use to set conditional probabilities")

    # Optional arguments
    parser.add_argument("--use-nc", "-nc",
        help="(For Chipseq only) Use the negative control mapped reads.",
        action="store_true")

    return parser.parse_args()

def main():
    args = parse_args()
    it_path = os.path.abspath( args.input_trace ) # use absolute path

    smo = StatmapOutput( sys.argv[1],
                         load_mapped_reads=True,
                         load_raw_reads=True,
                         load_nc=args.use_nc )

    # specify data structures to work with (using nc vs ip/normal)
    if args.use_nc:
        raw_rdb = smo.NC_rawread_db
        mpd_rdb = smo.NC_mpd_rdb
        cond_prbs_db = smo.NC_cond_prbs_db
    else:
        raw_rdb = smo.rawread_db
        mpd_rdb = smo.mpd_rdb
        cond_prbs_db = smo.cond_prbs_db

    # default not to reset conditional probabilities
    reset_cond_prbs = False

    trace = load_c_trace_from_file( it_path )

    # update the read conditional probabilities
    update_cond_prbs_from_trace_and_assay_type(
        mpd_rdb, cond_prbs_db, trace, smo.genome,
        smo.config.contents.assay_type )

    close_c_traces( trace )

    # write mapped reads to sam
    write_mapped_reads_to_sam(
        raw_rdb, mpd_rdb, cond_prbs_db, smo.genome,
        reset_cond_prbs, False,
        sys.stdout )

if __name__ == "__main__": main()

File: utilities/write_sam_from_mapped_reads_db_and_trace.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

#include "utility_common.h"

#include "../src/mapped_read.h"
#include "../src/genome.h"
#include "../src/sam.h"
#include "../src/rawread.h"
#include "../src/config_parsing.h"
#include "../src/fragment_length.h"
#include "../src/trace.h"
#include "../src/iterative_mapping.h"

/* make it possible to link */
int min_num_hq_bps = -1;
int num_threads = 1;

void usage()
{
    fprintf( stderr, "Usage: ./mapped_reads_to_sam output_directory input.bin.trace\n" );
}

int main( int argc, char** argv )
{
    if( argc != 3 )
    {
        usage();
        exit(1);
    }

    /* change to the output directory */
    int error = chdir( argv[1] );
    if( -1 == error )
    {
        perror( "FATAL       :  Cannot move into output directory ");
        exit( 1 );
    }
    
    /* load the configuration data */
    struct args_t* args;
    read_config_file_from_disk( &args );

    /* Load the genome */
    struct genome_data* genome;
    load_genome_from_disk( &genome, GENOME_FNAME );

    /* load the mapped read db */
    char* mpd_rd_fname = NULL;
    mpd_rd_fname = "mapped_reads.db";
    struct mapped_reads_db* mpd_rdb;
    open_mapped_reads_db( &mpd_rdb, mpd_rd_fname );

    struct cond_prbs_db_t* cond_prbs_db;
    init_cond_prbs_db_from_mpd_rdb( &cond_prbs_db, mpd_rdb );
    
    /* load the rawreads */
    struct rawread_db_t* raw_rdb;
    populate_rawread_db( &raw_rdb, "reads.unpaired", "reads.pair1", "reads.pair2" );
    
    if( NULL == raw_rdb )
    {
        fprintf( stderr, "FATAL       :  Can not load raw reads.\n" );
        exit( 1 );
    }    

    /* default not to reset conditional probabilities */
    enum bool reset_cond_prbs = false;
    
    /* load the trace that stores the marginal read 
       density that we are interested in */
    struct trace_t* traces;
    load_trace_from_file( &traces, argv[2] );

    /* update the read conditional probabilities based upon the assay
       and the correct trace */
    mmap_mapped_reads_db( mpd_rdb );
    index_mapped_reads_db( mpd_rdb );
    update_cond_prbs_from_trace_and_assay_type( 
        mpd_rdb, cond_prbs_db, traces, genome, args->assay_type );

    close_traces( traces );
        
    write_mapped_reads_to_sam( 
        raw_rdb, mpd_rdb, cond_prbs_db, genome, reset_cond_prbs, false, stdout );
    
    goto cleanup;
    
cleanup:
    free( args );
    free_genome( genome );
    close_mapped_reads_db( &mpd_rdb );
    close_rawread_db( raw_rdb );    

    return 0;
}


